### PR TITLE
[BugFix] Materialized view should update all its refreshed base table partitions after refresh

### PR DESCRIPTION
**Pull request by LiShuMing**
After https://github.com/StarRocks/starrocks/pull/26403/, we can prune int type partitions.
This will cause the following cases fail:
```
CREATE MATERIALIZED VIEW mv1 PARTITION BY k1 DISTRIBUTED BY HASH(k1) BUCKETS 10
REFRESH MANUAL AS SELECT k1, v1 as k2, v2 as k3 from t1;
                 
CREATE MATERIALIZED VIEW mv2 PARTITION BY k1 DISTRIBUTED BY HASH(k1) BUCKETS 10
REFRESH MANUAL AS SELECT k1, count(k2) as count_k2, sum(k3) as sum_k3 from mv1 group by k1;
                 
CREATE MATERIALIZED VIEW mv3 DISTRIBUTED BY HASH(k1)
REFRESH MANUAL AS SELECT k1, count_k2, sum_k3 from mv2 where k1 >1;
```
Changes:
- After mv refreshed, SR only updated base table partitions from plan which may be partition pruned.
This is not expected, I changed it to update its base table partitions if the mv has updated.
- For external ref table, add `externalRefBaseTableMVPartitionMap ` in taskRunContext;
-  Add `refBaseePartitionsToRefreshMap ` and `basePartitionsToRefreshMap ` in task_runs information_schema.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 169 / 206 (82.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|6|50.00%|[902, 903, 906]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|6|50.00%|[96, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|4|6|66.67%|[596, 603]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|9|11|81.82%|[1075, 2153]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|68|81|83.95%|[691, 715, 1157, 1161, 1189, 1190, 1215, 1217, 1220, 1225, 1233, 1256, 1263]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|65|78|83.33%|[220, 225, 226, 227, 346, 376, 1049, 1090, 1097, 1098, 1099, 1107, 1113]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|10|11|90.91%|[200]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 354 / 460 (76.96%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|3|6|50.00%|[902, 903, 906]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|33|51|64.71%|[308, 309, 396, 397, 400, 402, 403, 404, 405, 406, 407, 408, 409, 420, 422, 424, 456, 457]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|4|6|66.67%|[596, 603]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|2|3|66.67%|[149]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|181|237|76.37%|[219, 227, 376, 574, 579, 580, 582, 583, 584, 586, 591, 603, 617, 682, 896, 897, 898, 899, 900, 901, 902, 919, 925, 934, 951, 975, 979, 980, 981, 982, 985, 987, 988, 991, 992, 994, 995, 999, 1000, 1001, 1002, 1003, 1004, 1005, 1006, 1010, 1011, 1012, 1049, 1090, 1091, 1092, 1201, 1212, 1241, 1243]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|76|96|79.17%|[763, 764, 766, 767, 1215, 1217, 1225, 1231, 1233, 1254, 1255, 1256, 1263, 1264, 1275, 1304, 1323, 1324, 1325, 1326]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|19|22|86.36%|[96, 97, 98]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|10|11|90.91%|[200]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|9|10|90.00%|[2285]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|12|13|92.31%|[105]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 234 / 282 (82.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|11|27.27%|[87, 88, 89, 90, 91, 92, 93, 95]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|4|5|80.00%|[91]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|14|17|82.35%|[67, 111, 115]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|179|211|84.83%|[235, 236, 237, 244, 245, 377, 400, 655, 656, 665, 697, 698, 699, 797, 818, 959, 960, 1054, 1063, 1065, 1100, 1101, 1334, 1335, 1383, 1440, 1474, 1476, 1478, 1479, 1480, 1483]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|28|32|87.50%|[478, 480, 484, 486]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 236 / 282 (83.69%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|11|27.27%|[87, 88, 89, 90, 91, 92, 93, 95]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|4|5|80.00%|[91]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|14|17|82.35%|[67, 111, 115]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|181|211|85.78%|[235, 236, 237, 244, 245, 377, 400, 655, 656, 665, 697, 698, 699, 797, 818, 959, 960, 1054, 1063, 1065, 1334, 1335, 1383, 1440, 1474, 1476, 1478, 1479, 1480, 1483]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|28|32|87.50%|[478, 480, 484, 486]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 236 / 282 (83.69%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|11|27.27%|[87, 88, 89, 90, 91, 92, 93, 95]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|4|5|80.00%|[91]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|14|17|82.35%|[67, 111, 115]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|181|211|85.78%|[235, 236, 237, 244, 245, 377, 400, 655, 656, 665, 697, 698, 699, 797, 818, 959, 960, 1054, 1063, 1065, 1334, 1335, 1383, 1440, 1474, 1476, 1478, 1479, 1480, 1483]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|30|34|88.24%|[484, 486, 490, 492]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|

**Review by Astralidea — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java [hunk 1]
```diff
@@ -353,45 +367,81 @@ public static Set<String> getMVPartitionName(Table table, Column partitionColumn
...
+            List<String> partitionNameValues = toPartitionValues(partitionName);
+            PartitionKey partitionKey = createPartitionKey(
+                    ImmutableList.of(partitionNameValues.get(partitionColumnIndex)),
+                    ImmutableList.of(partitionColumns.get(partitionColumnIndex)),
+                    table.getType());
+            String mvPartitionName = generateMVPartitionName(partitionKey);
+            mvPartitionKeySetMap.computeIfAbsent(mvPartitionName, x -> Sets.newHashSet())
+                    .add(partitionName);
+        }
+        return mvPartitionKeySetMap;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:399]: Could you please add a MetaFunction to inspect this partition mapping  ?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:399]: Good advise~  I can add one function later.

In this pr, I have add  more  refresh partition info in `[MVTaskRunExtraMessage.java](https://github.com/StarRocks/starrocks/pull/28093/files/2d42190576e0d32d097f87352699e29d5d3f74cf#diff-2beea2501608af783672f127f5dc864766939d48a820104b4f71f0bb4440c371)`

File: fe/fe-core/src/main/java/com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java [hunk 2]
```diff
@@ -37,8 +37,14 @@ public class MVTaskRunExtraMessage implements Writable {
     private String partitionStart;
     @SerializedName("partitionEnd")
     private String partitionEnd;
+
+    // refreshed partitions of materialized view in this task run
     @SerializedName("mvPartitionsToRefresh")
     private Set<String> mvPartitionsToRefresh = Sets.newHashSet();
+    // refreshed partitions of the ref base table in this task run which should only have one table for now.
+    @SerializedName("refBaseePartitionsToRefreshMap")
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java:45]: fix it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 236 / 282 (83.69%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|11|27.27%|[87, 88, 89, 90, 91, 92, 93, 95]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|4|5|80.00%|[91]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|14|17|82.35%|[67, 111, 115]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|181|211|85.78%|[235, 236, 237, 244, 245, 377, 400, 655, 656, 665, 697, 698, 699, 797, 818, 959, 960, 1054, 1063, 1065, 1334, 1335, 1383, 1440, 1474, 1476, 1478, 1479, 1480, 1483]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|30|34|88.24%|[484, 486, 490, 492]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java [hunk 3]
```diff
@@ -1375,9 +1375,9 @@ private Set<String> getPartitionedMVRefreshPartitions() {
         mvPartitionNameToRangeMap.putAll(rangePartitionDiff.getAdds());
 
         Map<String, Set<String>> baseToMvNameRef = SyncPartitionUtils
-                .generatePartitionRefMap(basePartitionNameToRangeMap, mvPartitionNameToRangeMap);
+                .getIntersectedPartitions(basePartitionNameToRangeMap, mvPartitionNameToRangeMap);
         Map<String, Set<String>> mvToBaseNameRef = SyncPartitionUtils
-                .generatePartitionRefMap(mvPartitionNameToRangeMap, basePartitionNameToRangeMap);
+                .getIntersectedPartitions(mvPartitionNameToRangeMap, basePartitionNameToRangeMap);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:1380]: not a good name

**Review by ABingHuang — APPROVED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 234 / 282 (82.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|3|11|27.27%|[87, 88, 89, 90, 91, 92, 93, 95]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|4|5|80.00%|[91]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|14|17|82.35%|[67, 111, 115]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|179|211|84.83%|[235, 236, 237, 244, 245, 377, 400, 655, 656, 665, 697, 698, 699, 797, 818, 959, 960, 1054, 1063, 1065, 1100, 1101, 1334, 1335, 1383, 1440, 1474, 1476, 1478, 1479, 1480, 1483]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|30|34|88.24%|[484, 486, 490, 492]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**